iceoryx_hoofs/test/moduletests/test_posix_thread.cpp

```diff
@@ -35,66 +38,120 @@ class Thread_test : public Test
 
     void SetUp()
     {
-        m_run = true;
-        m_thread = new std::thread(&Thread_test::threadFunc, this);
     }
 
     void TearDown()
     {
-        m_run = false;
-        m_thread->join();
-        delete m_thread;
     }
 
     ~Thread_test()
     {
     }
 
-    void threadFunc()
-    {
-        while (m_run)
-        {
-            std::this_thread::sleep_for(std::chrono::milliseconds(10));
-        }
-    };
-
-    std::atomic_bool m_run{true};
-    std::thread* m_thread;
+    optional<Thread> sut;
 };
 
-#if !defined(__APPLE__)
+#if !defined(_WIN32) && !defined(__APPLE__)
+TEST_F(Thread_test, CreateThreadWithNonEmptyCallableSucceeds)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "0d1e439d-c84e-4a46-ac45-dc8be7530c32");
+    bool callableWasCalled = false;
+    Thread::callable_t callable = [&] { callableWasCalled = true; };
+    ASSERT_FALSE(ThreadBuilder().create(sut, callable).has_error());
+    sut.reset();
+    EXPECT_TRUE(callableWasCalled);
+}
+
+TEST_F(Thread_test, CreateThreadWithEmptyCallableFails)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "8058c282-ce33-42eb-80ed-4421ebac5652");
+    Thread::callable_t callable;
+    auto result = ThreadBuilder().create(sut, callable);
+    ASSERT_TRUE(result.has_error());
+    EXPECT_THAT(result.get_error(), Eq(ThreadError::EMPTY_CALLABLE));
+}
+
+TEST_F(Thread_test, DtorOfThreadBlocksUntilCallbackHasFinished)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "1062a036-e825-4f30-bfb8-00d5de47fdfd");
+
+    std::chrono::steady_clock::time_point start;
+    std::chrono::steady_clock::time_point end;
+    constexpr Duration TEST_WAIT_TIME = 10_ms;
+    Barrier threadSync;
+
+    {
+        optional<Thread> blockingThread;
+        ASSERT_FALSE(ThreadBuilder()
+                         .create(blockingThread,
+                                 [&] {
+                                     threadSync.wait();
+                                     end = std::chrono::steady_clock::now();
+                                 })
+                         .has_error());
+    }
+
+    ASSERT_FALSE(ThreadBuilder()
+                     .create(sut,
+                             [&] {
+                                 start = std::chrono::steady_clock::now();
+                                 std::this_thread::sleep_for(std::chrono::nanoseconds(TEST_WAIT_TIME.toNanoseconds()));
+                                 threadSync.notify();
+                             })
+                     .has_error());
+
```

Comment:
I think you forgot one line here, `blockingThread.reset()` since we would like to test if the dtor of the blocking thread blocks until the callback was finished.

But I may have an idea how to handle this all with only one thread:
```cxx
optional<Thread> blockingThread;
ASSERT_FALSE(ThreadBuilder()
                     .create(blockingThread,
                             [&] {
                                 threadSync.wait();
                                 std::this_thread::sleep_for(std::chrono::nanoseconds(TEST_WAIT_TIME.toNanoseconds()));
                             })
                     .has_error());

// since we want to ensure only roughly that the dtor blocks at least TEST_WAIT_TIME
// we can start the measurement already a little ahead right before we inform the thread and trigger the dtor
start = std::chrono::steady_clock::now(); 
threadSync.wait();
blockingThread.reset(); // destroy the object, ensure that this is blocking
end = std::chrono::steady_clock::now();

EXPECT_THAT(std::chrono::nanoseconds(end - start).count(), Gt(TEST_WAIT_TIME.toNanoseconds()));
```

Reply:
@elfenpiff You're right, I forgot the reset call but I replaced the test with your suggestion. Thanks!